cmd/juju/application/deploy.go

```diff
@@ -688,13 +691,62 @@ func (c *DeployCommand) deployCharm(
 			return errors.New("cannot use --num-units or --to with subordinate application")
 		}
 	}
-	serviceName := c.ApplicationName
-	if serviceName == "" {
-		serviceName = charmInfo.Meta.Name
+	applicationName := c.ApplicationName
+	if applicationName == "" {
+		applicationName = charmInfo.Meta.Name
 	}
+
+	// Process the --config args.
+	// We may have a single file arg specified, in which case
+	// it points to a YAML file keyed on the charm name and
+	// containing values for any charm settings.
+	// We may also have key/value pairs representing
+	// charm settings which overrides anything in the YAML file.
+	// If more than one file is specified, that is an error.
 	var configYAML []byte
-	if c.Config.Path != "" {
-		configYAML, err = c.Config.Read(ctx)
+	files, err := c.ConfigOptions.AbsoluteFileNames(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(files) > 1 {
+		return errors.Errorf("only a single config YAMl file can be specified, got %d", len(files))
+	}
+	if len(files) == 1 {
+		configYAML, err = ioutil.ReadFile(files[0])
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	attr, err := c.ConfigOptions.ReadAttrs(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	appConfig := make(map[string]string)
+	for k, v := range attr {
+		appConfig[k] = fmt.Sprintf("%v", v)
+	}
+
+	// Application facade V5 expects charm config to either all be in YAML
+	// or config map. If config map is specified, that overrides YAML.
+	// So we need to combine the two here to have only one.
+	if apiRoot.BestFacadeVersion("Application") < 6 && len(appConfig) > 0 {
+		var configFromFile map[string]map[interface{}]interface{}
+		err := yaml.Unmarshal(configYAML, &configFromFile)
+		if err != nil {
+			return errors.Annotate(err, "badly formatted YAMl config file")
+		}
+		if configFromFile == nil {
+			configFromFile = make(map[string]map[interface{}]interface{})
+		}
+		charmSettings, ok := configFromFile[applicationName]
+		if !ok {
+			charmSettings = make(map[interface{}]interface{})
+		}
+		for k, v := range appConfig {
+			charmSettings[k] = v
+		}
+		configFromFile[applicationName] = charmSettings
+		configYAML, err = yaml.Marshal(configFromFile)
```

Comment:
shouldn't you be setting appConfig to nil somewhere here?

Reply:
it doesn't matter because if there's a configYAML, it will mask any config map entries

Reply:
true, but it would be nicer - no unnecessary data on the wire

Reply:
done